File: backend/services/node_executors/prtg.py
"""
PRTG Node Executors

Workflow node executors for PRTG Network Monitor integration.
"""

import logging
from typing import Dict, Any

from backend.services.prtg_service import PRTGService

logger = logging.getLogger(__name__)


class PRTGGetDevicesExecutor:
    """Get devices from PRTG."""
    
    def execute(self, params: Dict[str, Any], context: Dict[str, Any]) -> Dict[str, Any]:
        try:
            service = PRTGService()
            
            devices = service.get_devices(
                group=params.get('group'),
                status=params.get('status'),
                search=params.get('search')
            )
            
            return {
                'success': True,
                'outputs': {
                    'devices': devices,
                    'count': len(devices)
                }
            }
        except Exception as e:
            logger.error(f"Error getting PRTG devices: {e}")
            return {
                'success': False,
                'error': str(e),
                'outputs': {
                    'devices': [],
                    'count': 0
                }
            }


class PRTGGetSensorsExecutor:
    """Get sensors from PRTG."""
    
    def execute(self, params: Dict[str, Any], context: Dict[str, Any]) -> Dict[str, Any]:
        try:
            service = PRTGService()
            
            device_id = params.get('device_id')
            if device_id:
                device_id = int(device_id)
            
            sensors = service.get_sensors(
                device_id=device_id,
                status=params.get('status'),
                sensor_type=params.get('sensor_type')
            )
            
            return {
                'success': True,
                'outputs': {
                    'sensors': sensors,
                    'count': len(sensors)
                }
            }
        except Exception as e:
            logger.error(f"Error getting PRTG sensors: {e}")
            return {
                'success': False,
                'error': str(e),
                'outputs': {
                    'sensors': [],
                    'count': 0
                }
            }


class PRTGGetSensorDetailsExecutor:
    """Get detailed sensor information."""
    
    def execute(self, params: Dict[str, Any], context: Dict[str, Any]) -> Dict[str, Any]:
        try:
            service = PRTGService()
            
            sensor_id = int(params.get('sensor_id'))
            sensor = service.get_sensor_details(sensor_id)
            
            return {
                'success': True,
                'outputs': {
                    'sensor': sensor,
                    'channels': sensor.get('channels', [])
                }
            }
        except Exception as e:
            logger.error(f"Error getting sensor details: {e}")
            return {
                'success': False,
                'error': str(e),
                'outputs': {
                    'sensor': None,
                    'channels': []
                }
            }


class PRTGGetAlertsExecutor:
    """Get current alerts from PRTG."""
    
    def execute(self, params: Dict[str, Any], context: Dict[str, Any]) -> Dict[str, Any]:
        try:
            service = PRTGService()
            
            alerts = service.get_alerts(status=params.get('status'))
            
            return {
                'success': True,
                'outputs': {
                    'alerts': alerts,
                    'count': len(alerts)
                }
            }
        except Exception as e:
            logger.error(f"Error getting PRTG alerts: {e}")
            return {
                'success': False,
                'error': str(e),
                'outputs': {
                    'alerts': [],
                    'count': 0
                }
            }


class PRTGGetGroupsExecutor:
    """Get device groups from PRTG."""
    
    def execute(self, params: Dict[str, Any], context: Dict[str, Any]) -> Dict[str, Any]:
        try:
            service = PRTGService()
            
            groups = service.get_groups()
            
            return {
                'success': True,
                'outputs': {
                    'groups': groups,
                    'count': len(groups)
                }
            }
        except Exception as e:
            logger.error(f"Error getting PRTG groups: {e}")
            return {
                'success': False,
                'error': str(e),
                'outputs': {
                    'groups': [],
                    'count': 0
                }
            }


class PRTGGetSensorHistoryExecutor:
    """Get historical data for a sensor."""
    
    def execute(self, params: Dict[str, Any], context: Dict[str, Any]) -> Dict[str, Any]:
        try:
            service = PRTGService()
            
            sensor_id = int(params.get('sensor_id'))
            result = service.get_sensor_history(
                sensor_id=sensor_id,
                start_date=params.get('start_date'),
                end_date=params.get('end_date'),
                avg=params.get('avg', 0)
            )
            
            return {
                'success': True,
                'outputs': {
                    'history': result.get('histdata', []),
                    'channels': result.get('channels', [])
                }
            }
        except Exception as e:
            logger.error(f"Error getting sensor history: {e}")
            return {
                'success': False,
                'error': str(e),
                'outputs': {
                    'history': [],
                    'channels': []
                }
            }


class PRTGAcknowledgeAlarmExecutor:
    """Acknowledge an alarm in PRTG."""
    
    def execute(self, params: Dict[str, Any], context: Dict[str, Any]) -> Dict[str, Any]:
        try:
            service = PRTGService()
            
            sensor_id = int(params.get('sensor_id'))
            message = params.get('message', 'Acknowledged by OpsConductor workflow')
            
            success = service.acknowledge_alarm(sensor_id, message)
            
            return {
                'success': success,
                'outputs': {
                    'success': success
                }
            }
        except Exception as e:
            logger.error(f"Error acknowledging alarm: {e}")
            return {
                'success': False,
                'error': str(e),
                'outputs': {
                    'success': False
                }
            }


class PRTGPauseObjectExecutor:
    """Pause a sensor, device, or group in PRTG."""
    
    def execute(self, params: Dict[str, Any], context: Dict[str, Any]) -> Dict[str, Any]:
        try:
            service = PRTGService()
            
            object_id = int(params.get('object_id'))
            duration = params.get('duration')
            if duration:
                duration = int(duration)
            message = params.get('message', 'Paused by OpsConductor workflow')
            
            success = service.pause_object(object_id, duration, message)
            
            return {
                'success': success,
                'outputs': {
                    'success': success
                }
            }
        except Exception as e:
            logger.error(f"Error pausing object: {e}")
            return {
                'success': False,
                'error': str(e),
                'outputs': {
                    'success': False
                }
            }


class PRTGResumeObjectExecutor:
    """Resume a paused object in PRTG."""
    
    def execute(self, params: Dict[str, Any], context: Dict[str, Any]) -> Dict[str, Any]:
        try:
            service = PRTGService()
            
            object_id = int(params.get('object_id'))
            success = service.resume_object(object_id)
            
            return {
                'success': success,
                'outputs': {
                    'success': success
                }
            }
        except Exception as e:
            logger.error(f"Error resuming object: {e}")
            return {
                'success': False,
                'error': str(e),
                'outputs': {
                    'success': False
                }
            }


class PRTGSyncToNetBoxExecutor:
    """Sync PRTG devices to NetBox."""
    
    def execute(self, params: Dict[str, Any], context: Dict[str, Any]) -> Dict[str, Any]:
        try:
            service = PRTGService()
            
            result = service.sync_to_netbox(
                dry_run=params.get('dry_run', True),
                device_ids=params.get('device_ids'),
                default_site=params.get('default_site'),
                default_role=params.get('default_role'),
                update_existing=params.get('update_existing', False),
                create_missing=params.get('create_missing', True)
            )
            
            return {
                'success': True,
                'outputs': {
                    'processed': result.get('processed', 0),
                    'created': result.get('created', 0),
                    'updated': result.get('updated', 0),
                    'skipped': result.get('skipped', 0),
                    'errors': result.get('errors', []),
                    'details': result.get('details', [])
                }
            }
        except Exception as e:
            logger.error(f"Error syncing to NetBox: {e}")
            return {
                'success': False,
                'error': str(e),
                'outputs': {
                    'processed': 0,
                    'created': 0,
                    'updated': 0,
                    'skipped': 0,
                    'errors': [{'error': str(e)}],
                    'details': []
                }
            }


class PRTGPreviewSyncExecutor:
    """Preview what would be synced to NetBox."""
    
    def execute(self, params: Dict[str, Any], context: Dict[str, Any]) -> Dict[str, Any]:
        try:
            service = PRTGService()
            
            result = service.preview_netbox_sync()
            
            return {
                'success': True,
                'outputs': {
                    'total_prtg_devices': result.get('total_prtg_devices', 0),
                    'existing_in_netbox': result.get('existing_in_netbox', 0),
                    'to_create': result.get('to_create', 0),
                    'devices_to_create': result.get('devices_to_create', []),
                    'devices_existing': result.get('devices_existing', [])
                }
            }
        except Exception as e:
            logger.error(f"Error previewing sync: {e}")
            return {
                'success': False,
                'error': str(e),
                'outputs': {
                    'total_prtg_devices': 0,
                    'existing_in_netbox': 0,
                    'to_create': 0,
                    'devices_to_create': [],
                    'devices_existing': []
                }
            }


# Executor registry for PRTG nodes
PRTG_EXECUTORS = {
    'prtg.get_devices': PRTGGetDevicesExecutor(),
    'prtg.get_sensors': PRTGGetSensorsExecutor(),
    'prtg.get_sensor_details': PRTGGetSensorDetailsExecutor(),
    'prtg.get_alerts': PRTGGetAlertsExecutor(),
    'prtg.get_groups': PRTGGetGroupsExecutor(),
    'prtg.get_sensor_history': PRTGGetSensorHistoryExecutor(),
    'prtg.acknowledge_alarm': PRTGAcknowledgeAlarmExecutor(),
    'prtg.pause_object': PRTGPauseObjectExecutor(),
    'prtg.resume_object': PRTGResumeObjectExecutor(),
    'prtg.sync_to_netbox': PRTGSyncToNetBoxExecutor(),
    'prtg.preview_sync': PRTGPreviewSyncExecutor(),
}

File: _archive/legacy_frontend/frontend/src/features/workflow-builder/packages/prtg.js
/**
 * PRTG Node Package
 * 
 * Nodes for integrating with PRTG Network Monitor:
 * - Get devices and sensors
 * - Query alerts
 * - Acknowledge/pause/resume objects
 * - Sync to NetBox
 */

const prtgPackage = {
  id: 'prtg',
  name: 'PRTG Network Monitor',
  description: 'Integration with PRTG Network Monitor for device monitoring and alerts',
  version: '1.0.0',
  icon: 'Activity',
  color: '#00a651',
  
  nodes: {
    // ========================================================================
    // Query Nodes
    // ========================================================================
    
    'prtg:get-devices': {
      id: 'prtg:get-devices',
      name: 'Get PRTG Devices',
      description: 'Retrieve devices from PRTG',
      category: 'query',
      icon: 'Server',
      color: '#00a651',
      
      parameters: [
        {
          id: 'group',
          label: 'Group Filter',
          type: 'string',
          required: false,
          placeholder: 'Network Devices',
          description: 'Filter by group name (partial match)'
        },
        {
          id: 'status',
          label: 'Status Filter',
          type: 'select',
          required: false,
          options: [
            { value: '', label: 'All Statuses' },
            { value: 'up', label: 'Up' },
            { value: 'down', label: 'Down' },
            { value: 'warning', label: 'Warning' },
            { value: 'paused', label: 'Paused' }
          ],
          description: 'Filter by device status'
        },
        {
          id: 'search',
          label: 'Search',
          type: 'string',
          required: false,
          placeholder: 'Search term',
          description: 'Search by device name or IP'
        }
      ],
      
      outputs: [
        { id: 'devices', label: 'Devices', type: 'array' },
        { id: 'count', label: 'Count', type: 'number' }
      ],
      
      executor: 'prtg.get_devices'
    },
    
    'prtg:get-sensors': {
      id: 'prtg:get-sensors',
      name: 'Get PRTG Sensors',
      description: 'Retrieve sensors from PRTG',
      category: 'query',
      icon: 'Gauge',
      color: '#00a651',
      
      parameters: [
        {
          id: 'device_id',
          label: 'Device ID',
          type: 'string',
          required: false,
          placeholder: '1234',
          description: 'Filter by PRTG device ID'
        },
        {
          id: 'status',
          label: 'Status Filter',
          type: 'select',
          required: false,
          options: [
            { value: '', label: 'All Statuses' },
            { value: 'up', label: 'Up' },
            { value: 'down', label: 'Down' },
            { value: 'warning', label: 'Warning' },
            { value: 'unusual', label: 'Unusual' },
            { value: 'paused', label: 'Paused' }
          ]
        },
        {
          id: 'sensor_type',
          label: 'Sensor Type',
          type: 'string',
          required: false,
          placeholder: 'ping',
          description: 'Filter by sensor type'
        }
      ],
      
      outputs: [
        { id: 'sensors', label: 'Sensors', type: 'array' },
        { id: 'count', label: 'Count', type: 'number' }
      ],
      
      executor: 'prtg.get_sensors'
    },
    
    'prtg:get-sensor-details': {
      id: 'prtg:get-sensor-details',
      name: 'Get Sensor Details',
      description: 'Get detailed information for a specific sensor',
      category: 'query',
      icon: 'Info',
      color: '#00a651',
      
      parameters: [
        {
          id: 'sensor_id',
          label: 'Sensor ID',
          type: 'string',
          required: true,
          placeholder: '1234',
          description: 'PRTG sensor ID'
        }
      ],
      
      inputs: [
        { id: 'sensor_id', label: 'Sensor ID', type: 'string' }
      ],
      
      outputs: [
        { id: 'sensor', label: 'Sensor Details', type: 'object' },
        { id: 'channels', label: 'Channels', type: 'array' }
      ],
      
      executor: 'prtg.get_sensor_details'
    },
    
    'prtg:get-alerts': {
      id: 'prtg:get-alerts',
      name: 'Get PRTG Alerts',
      description: 'Get current alerts from PRTG',
      category: 'query',
      icon: 'AlertTriangle',
      color: '#ef4444',
      
      parameters: [
        {
          id: 'status',
          label: 'Alert Status',
          type: 'select',
          required: false,
          options: [
            { value: '', label: 'All Alerts' },
            { value: 'down', label: 'Down' },
            { value: 'warning', label: 'Warning' },
            { value: 'unusual', label: 'Unusual' }
          ]
        }
      ],
      
      outputs: [
        { id: 'alerts', label: 'Alerts', type: 'array' },
        { id: 'count', label: 'Count', type: 'number' }
      ],
      
      executor: 'prtg.get_alerts'
    },
    
    'prtg:get-groups': {
      id: 'prtg:get-groups',
      name: 'Get PRTG Groups',
      description: 'Get device groups from PRTG',
      category: 'query',
      icon: 'Folder',
      color: '#00a651',
      
      parameters: [],
      
      outputs: [
        { id: 'groups', label: 'Groups', type: 'array' },
        { id: 'count', label: 'Count', type: 'number' }
      ],
      
      executor: 'prtg.get_groups'
    },
    
    'prtg:get-sensor-history': {
      id: 'prtg:get-sensor-history',
      name: 'Get Sensor History',
      description: 'Get historical data for a sensor',
      category: 'query',
      icon: 'TrendingUp',
      color: '#00a651',
      
      parameters: [
        {
          id: 'sensor_id',
          label: 'Sensor ID',
          type: 'string',
          required: true,
          placeholder: '1234'
        },
        {
          id: 'start_date',
          label: 'Start Date',
          type: 'string',
          required: false,
          placeholder: '2024-01-01-00-00-00',
          description: 'Format: YYYY-MM-DD-HH-MM-SS'
        },
        {
          id: 'end_date',
          label: 'End Date',
          type: 'string',
          required: false,
          placeholder: '2024-01-02-00-00-00'
        },
        {
          id: 'avg',
          label: 'Averaging (seconds)',
          type: 'number',
          required: false,
          default: 0,
          description: '0 = no averaging'
        }
      ],
      
      inputs: [
        { id: 'sensor_id', label: 'Sensor ID', type: 'string' }
      ],
      
      outputs: [
        { id: 'history', label: 'History Data', type: 'array' },
        { id: 'channels', label: 'Channels', type: 'array' }
      ],
      
      executor: 'prtg.get_sensor_history'
    },
    
    // ========================================================================
    // Action Nodes
    // ========================================================================
    
    'prtg:acknowledge-alarm': {
      id: 'prtg:acknowledge-alarm',
      name: 'Acknowledge Alarm',
      description: 'Acknowledge an alarm in PRTG',
      category: 'action',
      icon: 'CheckCircle',
      color: '#10b981',
      
      parameters: [
        {
          id: 'sensor_id',
          label: 'Sensor ID',
          type: 'string',
          required: true,
          placeholder: '1234'
        },
        {
          id: 'message',
          label: 'Acknowledgment Message',
          type: 'textarea',
          required: false,
          placeholder: 'Acknowledged by OpsConductor workflow'
        }
      ],
      
      inputs: [
        { id: 'sensor_id', label: 'Sensor ID', type: 'string' }
      ],
      
      outputs: [
        { id: 'success', label: 'Success', type: 'boolean' }
      ],
      
      executor: 'prtg.acknowledge_alarm'
    },
    
    'prtg:pause-object': {
      id: 'prtg:pause-object',
      name: 'Pause Object',
      description: 'Pause a sensor, device, or group in PRTG',
      category: 'action',
      icon: 'PauseCircle',
      color: '#f59e0b',
      
      parameters: [
        {
          id: 'object_id',
          label: 'Object ID',
          type: 'string',
          required: true,
          placeholder: '1234',
          description: 'Sensor, device, or group ID'
        },
        {
          id: 'duration',
          label: 'Duration (minutes)',
          type: 'number',
          required: false,
          placeholder: '60',
          description: 'Leave empty for indefinite pause'
        },
        {
          id: 'message',
          label: 'Pause Message',
          type: 'textarea',
          required: false,
          placeholder: 'Paused by OpsConductor workflow'
        }
      ],
      
      inputs: [
        { id: 'object_id', label: 'Object ID', type: 'string' }
      ],
      
      outputs: [
        { id: 'success', label: 'Success', type: 'boolean' }
      ],
      
      executor: 'prtg.pause_object'
    },
    
    'prtg:resume-object': {
      id: 'prtg:resume-object',
      name: 'Resume Object',
      description: 'Resume a paused sensor, device, or group',
      category: 'action',
      icon: 'PlayCircle',
      color: '#10b981',
      
      parameters: [
        {
          id: 'object_id',
          label: 'Object ID',
          type: 'string',
          required: true,
          placeholder: '1234'
        }
      ],
      
      inputs: [
        { id: 'object_id', label: 'Object ID', type: 'string' }
      ],
      
      outputs: [
        { id: 'success', label: 'Success', type: 'boolean' }
      ],
      
      executor: 'prtg.resume_object'
    },
    
    // ========================================================================
    // Sync Nodes
    // ========================================================================
    
    'prtg:sync-to-netbox': {
      id: 'prtg:sync-to-netbox',
      name: 'Sync to NetBox',
      description: 'Sync PRTG devices to NetBox',
      category: 'sync',
      icon: 'RefreshCw',
      color: '#3b82f6',
      
      parameters: [
        {
          id: 'dry_run',
          label: 'Dry Run',
          type: 'boolean',
          default: true,
          description: 'Preview changes without making them'
        },
        {
          id: 'create_missing',
          label: 'Create Missing Devices',
          type: 'boolean',
          default: true,
          description: 'Create devices in NetBox that exist in PRTG'
        },
        {
          id: 'update_existing',
          label: 'Update Existing Devices',
          type: 'boolean',
          default: false,
          description: 'Update devices that already exist in NetBox'
        },
        {
          id: 'default_site',
          label: 'Default Site ID',
          type: 'number',
          required: false,
          description: 'NetBox site ID for new devices'
        },
        {
          id: 'default_role',
          label: 'Default Role ID',
          type: 'number',
          required: false,
          description: 'NetBox device role ID for new devices'
        }
      ],
      
      outputs: [
        { id: 'processed', label: 'Processed', type: 'number' },
        { id: 'created', label: 'Created', type: 'number' },
        { id: 'updated', label: 'Updated', type: 'number' },
        { id: 'skipped', label: 'Skipped', type: 'number' },
        { id: 'errors', label: 'Errors', type: 'array' },
        { id: 'details', label: 'Details', type: 'array' }
      ],
      
      executor: 'prtg.sync_to_netbox'
    },
    
    'prtg:preview-sync': {
      id: 'prtg:preview-sync',
      name: 'Preview NetBox Sync',
      description: 'Preview what would be synced to NetBox',
      category: 'sync',
      icon: 'Eye',
      color: '#6366f1',
      
      parameters: [],
      
      outputs: [
        { id: 'total_prtg_devices', label: 'Total PRTG Devices', type: 'number' },
        { id: 'existing_in_netbox', label: 'Existing in NetBox', type: 'number' },
        { id: 'to_create', label: 'To Create', type: 'number' },
        { id: 'devices_to_create', label: 'Devices to Create', type: 'array' },
        { id: 'devices_existing', label: 'Existing Devices', type: 'array' }
      ],
      
      executor: 'prtg.preview_sync'
    },
    
    // ========================================================================
    // Trigger Nodes
    // ========================================================================
    
    'prtg:alert-trigger': {
      id: 'prtg:alert-trigger',
      name: 'PRTG Alert Trigger',
      description: 'Trigger workflow when PRTG alert is received via webhook',
      category: 'trigger',
      icon: 'Zap',
      color: '#ef4444',
      
      parameters: [
        {
          id: 'severity_filter',
          label: 'Severity Filter',
          type: 'multiselect',
          required: false,
          options: [
            { value: 'down', label: 'Down' },
            { value: 'warning', label: 'Warning' },
            { value: 'unusual', label: 'Unusual' }
          ],
          description: 'Only trigger for these severities (empty = all)'
        },
        {
          id: 'device_filter',
          label: 'Device Name Filter',
          type: 'string',
          required: false,
          placeholder: 'router-*',
          description: 'Glob pattern to match device names'
        },
        {
          id: 'sensor_filter',
          label: 'Sensor Name Filter',
          type: 'string',
          required: false,
          placeholder: '*ping*',
          description: 'Glob pattern to match sensor names'
        }
      ],
      
      outputs: [
        { id: 'alert', label: 'Alert Data', type: 'object' },
        { id: 'device_name', label: 'Device Name', type: 'string' },
        { id: 'sensor_name', label: 'Sensor Name', type: 'string' },
        { id: 'severity', label: 'Severity', type: 'string' },
        { id: 'message', label: 'Message', type: 'string' },
        { id: 'host', label: 'Host IP', type: 'string' }
      ],
      
      executor: 'prtg.alert_trigger'
    }
  },
  
  categories: {
    trigger: {
      label: 'Triggers',
      icon: 'Zap',
      order: 1
    },
    query: {
      label: 'Query',
      icon: 'Search',
      order: 2
    },
    action: {
      label: 'Actions',
      icon: 'Play',
      order: 3
    },
    sync: {
      label: 'Sync',
      icon: 'RefreshCw',
      order: 4
    }
  }
};

export default prtgPackage;
